lib/srv/db/mysql/autousers.go

```diff
@@ -0,0 +1,367 @@
+	"context"
+	_ "embed"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"hash/fnv"
+	"strings"
+
+	"github.com/go-mysql-org/go-mysql/client"
+	"github.com/gravitational/trace"
+	"golang.org/x/exp/slices"
+
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/srv/db/common"
+)
+
+// activateUserDetails contains details about the user activation request and
+// will be marshaled into a JSON parameter for the stored procedure.
+type activateUserDetails struct {
+	// Roles is a list of roles to be assigned to the user.
+	//
+	// MySQL stored procedure does not accept array of strings thus using a
+	// JSON to bypass this limit.
+	Roles []string `json:"roles"`
+	// AuthOptions specifies auth options like "IDENTIFIED xxx" used when
+	// creating a new user.
+	//
+	// Using a JSON string can bypass VARCHAR character limit.
+	AuthOptions string `json:"auth_options"`
+	// Attributes specifies user attributes used when creating a new user.
+	//
+	// User attributes is a MySQL JSON in MySQL databases.
+	//
+	// To check current user's attribute:
+	// SELECT * FROM INFORMATION_SCHEMA.USER_ATTRIBUTES WHERE CONCAT(USER, '@', HOST) = current_user()
+	//
+	// Reference:
+	// https://dev.mysql.com/doc/refman/8.0/en/information-schema-user-attributes-table.html
+	Attributes struct {
+		// User is the original Teleport user name.
+		//
+		// Find a Teleport user (with "admin" privilege):
+		// SELECT * FROM INFORMATION_SCHEMA.USER_ATTRIBUTES WHERE ATTRIBUTE->"$.user" = "teleport-user-name";
+		User string `json:"user"`
+	} `json:"attributes"`
+}
+
+// ActivateUser creates or enables the database user.
+func (e *Engine) ActivateUser(ctx context.Context, sessionCtx *common.Session) error {
+	if sessionCtx.Database.GetAdminUser() == "" {
+		return trace.BadParameter("Teleport does not have admin user configured for this database")
+	}
+
+	conn, err := e.connectAsAdminUser(ctx, sessionCtx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer conn.Close()
+
+	// Setup "teleport-auto-user" and stored procedures.
+	if err := e.setupDatabaseForAutoUsers(conn, sessionCtx); err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Use "teleport-<hash>" in case DatabaseUser is over max username length.
+	sessionCtx.DatabaseUser = maybeHashUsername(sessionCtx.DatabaseUser, maxUsernameLength(conn))
+
+	e.Log.Infof("Activating MySQL user %q with roles %v for %v.", sessionCtx.DatabaseUser, sessionCtx.DatabaseRoles, sessionCtx.Identity.Username)
+
+	details, err := makeActivateUserDetails(sessionCtx, sessionCtx.Identity.Username)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	callCommand := fmt.Sprintf("CALL %s('%s', '%s')", activateUserProcedureName, sessionCtx.DatabaseUser, details)
+
+	_, err = conn.Execute(callCommand)
+	if err != nil {
+		if strings.Contains(err.Error(), "Operation CREATE USER failed") {
+			return trace.AlreadyExists("user %q already exists in this MySQL database and is not managed by Teleport", sessionCtx.DatabaseUser)
+		}
```

Comment:
Are we guessing that pre-existing user was the cause here? It might be a good idea to log the raw error in case it is useful for debugging.